01/19~01/22
4. breast_cancer 데이터 활용 실습 진행 
=> KNN, Decision Tree, Random Forest 이외에 다른 분류 이용해보거나 조사, 해당 기법들 각각 설명, 결과 설명 등
5. 3, 4에서 만들었던 모델의 성능을 높이는 방법을 각자 희망하는대로 하여 성능을 높이고 그 내용과 코드를 주석과 마크다운을 이용해서 정리하기

In [1]:
import pandas as pd
from sklearn.datasets import load_breast_cancer

In [11]:
cancer = load_breast_cancer()

In [3]:
cancer

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
         1.189e-01],
        [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
         8.902e-02],
        [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
         8.758e-02],
        ...,
        [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
         7.820e-02],
        [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
         1.240e-01],
        [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
         7.039e-02]]),
 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
        0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
        1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
        1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
        1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0

In [18]:
# 로드한 cancer 전체 데이터에 key 값을 출력
print(cancer.keys())
# cancer 전체 데이터 중 data에 대한 전체 행, 열 길이를 출력
print(cancer.data.shape)
# cancer 데이터에 컬럼 이름을 출력 
print(cancer.feature_names)
# cancer 데이터에 타겟 이름을 출력 
print(cancer.target_names)

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
(569, 30)
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
['malignant' 'benign']


In [25]:
cancer.data.shape
#데이터셋에 총 569개의 데이터가 있음

(569, 30)

In [21]:
cancer.target.sum()
#그 중 양성 데이터 개수는 357

357

In [20]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test=train_test_split(cancer.data, cancer.target, stratify=cancer.target, random_state=0)

# KNN

#### 매개변수 n_neighbors(근접한 이웃의 수)에 따라 타겟을 분류 -> 매개변수가 1이면 해당 데이터의 클래스를 그대로 따름, 2 이상이면 이웃 데이터의 클래스들 중 더 많은 클래스에 속함
#### 단순한 모델로 실제 분석에서는 사용을 거의 하지 않는다
#### k가 너무 크면 과소적합, 신중한 모델이 아닌 예측 결과가 모두 비슷해진다
#### k가 너무 작으면 과적합, 이웃에 너무 민감해진다

In [38]:
from sklearn.neighbors import KNeighborsClassifier

In [39]:
knn = KNeighborsClassifier()

In [40]:
knn.fit(X_train, y_train)
print("Accuracy on training set: {:.3f}".format(knn.score(X_train, y_train))) 
print("Accuracy on test set: {:.3f}".format(knn.score(X_test, y_test)))

Accuracy on training set: 0.946
Accuracy on test set: 0.916


정확도: 약 0.9

# Decision Tree

#### 나무가 뒤집어있는 모양의 모델로 가장 타겟을 잘 분류할 수 있는 피처에서 기준을 정해 밑으로 점점 깊어지는 모델
#### 따라서 첫 기준에 쓰이는 피처가 가장 타겟을 효과적으로 분류하는 피처
#### 정확도가 높고 이해하기 쉬운 것이 특징이나 주어진 훈련 데이터셋에 과적합될 가능성이 높음

In [37]:
from sklearn.tree import DecisionTreeClassifier

In [41]:
tree = DecisionTreeClassifier(random_state=0)
tree.fit(X_train, y_train)
print("Accuracy on training set: {:.3f}".format(tree.score(X_train, y_train))) 
print("Accuracy on test set: {:.3f}".format(tree.score(X_test, y_test)))

Accuracy on training set: 1.000
Accuracy on test set: 0.902


정확도: 약 0.9
정확도에서 볼 수 있듯이 훈련 데이터셋에서의 정확도는 1로 과적합되어있다

# Randomforest

#### 각기 다른 나무들의 모음
#### 과적합 가능성이 큰 decision tree를 서로 다른 모양으로 많이 만들어서 평균화한다 -> 과적합 양을 줄인다

In [42]:
from sklearn.ensemble import RandomForestClassifier

In [43]:
forest=RandomForestClassifier(random_state=0)
forest.fit(X_train, y_train)
print("Accuracy on training set: {:.3f}".format(forest.score(X_train, y_train))) 
print("Accuracy on test set: {:.3f}".format(forest.score(X_test, y_test)))

Accuracy on training set: 0.998
Accuracy on test set: 0.944


정확도: 약 0.94
decision tree보다 과적합 양이 줄어들어 향상된 정확도를 만들 수 있다.

# Gradient Boosting

#### 많은 양의 트리를 평균화하는 것이 아닌 각 트리가 앞서 만들어진 트리의 실수를 수정하는 직렬 방식
#### 깊이가 얕은 모델의 결합

In [44]:
from sklearn.ensemble import GradientBoostingClassifier

In [45]:
gbrt = GradientBoostingClassifier(random_state=0)
gbrt.fit(X_train, y_train)
print("Training set accuracy: {:.3f}".format(gbrt.score(X_train, y_train)))
print("Test set accuracy: {:.3f}".format(gbrt.score(X_test, y_test)))

Training set accuracy: 1.000
Test set accuracy: 0.958


정확도: 약 0.96 
마찬가지로 decision tree보다 과적합 양이 줄어들어 향상된 정확도를 만들 수 있다.

## 실제로 가장 널리 사용되는 알고리즘이 바로 randomforest와 gradientboosting